### PR TITLE
feat: add accounts for dockerhub from pipeline

### PR DESCRIPTION
**Pull request by smerle33**
add a withDockerPushCredentials and withDockerPullCredentials function to help dispatch connexion to dockerhub and avoid rate limit

as per : https://github.com/jenkins-infra/helpdesk/issues/2837

### CODE REVIEW

**Comment by smerle33** [comment]
still to do : 

- [ ]  create unit test specific to those 2 new functions
- [ ] create those accounts on dockerhub and add them in the jcasc (with sops)

File: test/groovy/BuildPluginStepTests.groovy [hunk 1]
```diff
@@ -77,7 +77,7 @@ class BuildPluginStepTests extends BaseTest {
     def script = loadScript(scriptName)
     def configurations = script.getConfigurations(platforms: ['bar', 'foo'])
     println configurations
-    assertEquals(configurations.size, 2)
+    assertEquals(configurations.size(), 2)
```
Comment (dduportal) [inline test/groovy/BuildPluginStepTests.groovy:80]: what is the reason of this change?

Comment (smerle33) [reply test/groovy/BuildPluginStepTests.groovy:80]: the unit test on that part of the code was broken ... missing those 2 () ... I discovered this by chance and took the opportunity to correct it ... part of why this is in draft, I planned to discuss this with you tomorrow.

Comment (dduportal) [reply test/groovy/BuildPluginStepTests.groovy:80]: As discussed together:
- The unit tests are failing when Java 17 is used. Since the requirement here is Java 11, this change is out of scope for this PR
- This change (`s/size/size()`)  is correct and easier to read for non Groovy specialist: might be worth it to add it on another PR, along with enabling the skipped test (or deleting it)

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 2]
```diff
@@ -1,22 +1,37 @@
...
   String jenkinsURL
+  String jenkinsHostname
 
   public InfraConfig(Object env) {
     this.jenkinsURL = env?.JENKINS_URL ?: ''
+    if (this.jenkinsURL != '') {
+      this.jenkinsHostname = new URL(this.jenkinsURL).getHost()
+    } else {
+      this.jenkinsHostname =  ''
+    }
```
Comment (lemeurherve) [inline src/io/jenkins/infra/InfraConfig.groovy:14]:
```suggestion
    this.jenkinsHostname =  ''
    if (this.jenkinsURL != '') {
      this.jenkinsHostname = new URL(this.jenkinsURL).getHost()
    }
```
(nit not blocking)

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 3]
```diff
@@ -1,22 +1,37 @@
...
   }
 
-  Boolean isRunningOnJenkinsInfra() {
-    return jenkinsURL == 'https://ci.jenkins.io/' || isTrusted() || isInfra()
+  Boolean isJenkinsURLcontains(String search) {
+    return jenkinsURL.startsWith('https://'+search+'.jenkins.io/') || jenkinsURL.startsWith('https://'+search+'.jenkins.io:')
```
Comment (lemeurherve) [inline src/io/jenkins/infra/InfraConfig.groovy:18]:
```suggestion
    return jenkinsURL.startsWith('https://' + search + '.jenkins.io/') || jenkinsURL.startsWith('https://' + search + '.jenkins.io:')
```

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 4]
```diff
@@ -27,4 +42,36 @@ class InfraConfig implements Serializable {
...
+        return [error:false, organisation:"cijenkinsioinfra", credentialId:"cijenkinsioinfra-dockerhub-pull"]
+        break;
+      case "trusted.ci.jenkins.io": 
+        return [error:false, organisation:"trustedcijenkinsio", credentialId:"trustedcijenkinsio-dockerhub-pull"]
+        break;
+      case "infra.ci.jenkins.io": 
+        return [error:false, organisation:"infracijenkinsio", credentialId:"infracijenkinsio-dockerhub-pull"]
+        break;
+    }
+    return [error:true, msg:"Cannot use Docker credentials outside of jenkins infra environment"]
```
Comment (lemeurherve) [inline src/io/jenkins/infra/InfraConfig.groovy:59]:
```suggestion
        return [error: false, organisation: "cijenkinsioinfra", credentialId: "cijenkinsioinfra-dockerhub-pull"]
        break;
      case "trusted.ci.jenkins.io": 
        return [error: false, organisation: "trustedcijenkinsio", credentialId: "trustedcijenkinsio-dockerhub-pull"]
        break;
      case "infra.ci.jenkins.io": 
        return [error: false, organisation: "infracijenkinsio", credentialId: "infracijenkinsio-dockerhub-pull"]
        break;
    }
    return [error: true, msg: "Cannot use Docker credentials outside of jenkins infra environment"]
```
(nit)

I would have put line 59 as the `default` switch case.

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 5]
```diff
@@ -27,4 +42,36 @@ class InfraConfig implements Serializable {
...
+        return [error:false, organisation:"jenkins4eval", credentialId:"jenkins4eval-dockerhub-push"]
+        break;
+      case "trusted.ci.jenkins.io": 
+        return [error:false, organisation:"jenkins", credentialId:"jenkins-dockerhub-push"]
+        break;
+      case "infra.ci.jenkins.io": 
+        return [error:false, organisation:"jenkinsciinfra", credentialId:"jenkinsciinfra-dockerhub-push"]
+        break;
+    }
+    return [error:true, msg:"Cannot use Docker credentials outside of jenkins infra environment"]
```
Comment (lemeurherve) [inline src/io/jenkins/infra/InfraConfig.groovy:75]:
```suggestion
      case "ci.jenkins.io": 
        return [error: false, organisation: "jenkins4eval", credentialId: "jenkins4eval-dockerhub-push"]
        break;
      case "trusted.ci.jenkins.io": 
        return [error: false, organisation: "jenkins", credentialId: "jenkins-dockerhub-push"]
        break;
      case "infra.ci.jenkins.io": 
        return [error: false, organisation: "jenkinsciinfra", credentialId: "jenkinsciinfra-dockerhub-push"]
        break;
    }
    return [error: true, msg: "Cannot use Docker credentials outside of jenkins infra environment"]
```
Same comment as above

File: test/groovy/InfraStepTests.groovy [hunk 6]
```diff
@@ -38,6 +38,8 @@ class InfraStepTests extends BaseTest {
     script.withDockerCredentials() {
       isOK = true
     }
+    print "as it been executed within the dockers credentials ? "
```
Comment (lemeurherve) [inline test/groovy/InfraStepTests.groovy:41]:
```suggestion
    print "Has it been executed within the Docker credentials?"
```

**Review by lemeurherve — DISMISSED** [review]
Great job!

File: test/groovy/InfraStepTests.groovy (hunk 6 above)
Comment (smerle33) [reply test/groovy/InfraStepTests.groovy:41]: 👍 no this should be deleted ... sorry a debug leftover ;)

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 7]
```diff
@@ -27,4 +41,36 @@ class InfraConfig implements Serializable {
...
+        return [error: false, organisation: "jenkins4eval", credentialId: "jenkins4eval-dockerhub-push"]
+        break;
+      case "trusted.ci.jenkins.io": 
+        return [error: false, organisation: "jenkins", credentialId: "jenkins-dockerhub-push"]
+        break;
+      case "infra.ci.jenkins.io": 
+        return [error: false, organisation: "jenkinsciinfra", credentialId: "jenkinsciinfra-dockerhub-push"]
+        break;
+    }
+    return [error: true, msg: "Cannot use Docker credentials outside of jenkins infra environment"]
```
Comment (lemeurherve) [inline src/io/jenkins/infra/InfraConfig.groovy:74]:
```suggestion
      case "infra.ci.jenkins.io": 
        return [error: false, organisation: "infracijenkinsio", credentialId: "infracijenkinsio-dockerhub-pull"]
        break;
      default:
        return [error: true, msg: "Cannot use Docker credentials outside of jenkins infra environment"]
    }
  }

  // Returns the Docker Informations for pushing images
  def getDockerPushOrgAndCredentialsId() {
    switch(jenkinsHostname){
      case "ci.jenkins.io": 
        return [error: false, organisation: "jenkins4eval", credentialId: "jenkins4eval-dockerhub-push"]
        break;
      case "trusted.ci.jenkins.io": 
        return [error: false, organisation: "jenkins", credentialId: "jenkins-dockerhub-push"]
        break;
      case "infra.ci.jenkins.io": 
        return [error: false, organisation: "jenkinsciinfra", credentialId: "jenkinsciinfra-dockerhub-push"]
        break;
      default:
        return [error: true, msg: "Cannot use Docker credentials outside of jenkins infra environment"]
    }
```

File: test/groovy/InfraStepTests.groovy [hunk 8]
```diff
@@ -57,6 +57,61 @@ class InfraStepTests extends BaseTest {
...
+  void testWithDockerPullCredentialsOutsideInfra() throws Exception {
+    def script = loadScript(scriptName)
+    env.JENKINS_URL = 'https://foo/'
+    def isOK = false
+    script.withDockerPullCredentials() {
+      isOK = true
+    }
+    printCallStack()
+    assertFalse(isOK)
+    assertTrue(assertMethodCallContainsPattern('echo', 'Cannot use Docker credentials outside of jenkins infra environment'))
```
Comment (lemeurherve) [inline test/groovy/InfraStepTests.groovy:110]:
```suggestion
    assertTrue(assertMethodCallContainsPattern('echo', 'Cannot use Docker credentials outside of jenkins infra environments'))
    assertJobStatusSuccess()
  }

  @Test
  void testWithDockerPullCredentials() throws Exception {
    def script = loadScript(scriptName)
    env.JENKINS_URL = 'https://ci.jenkins.io/'
    def isOK = false
    script.withDockerPullCredentials() {
      isOK = true
    }
    printCallStack()
    assertTrue(isOK)
    assertJobStatusSuccess()
  }

  @Test
  void testWithDockerPullCredentialsOutsideInfra() throws Exception {
    def script = loadScript(scriptName)
    env.JENKINS_URL = 'https://foo/'
    def isOK = false
    script.withDockerPullCredentials() {
      isOK = true
    }
    printCallStack()
    assertFalse(isOK)
    assertTrue(assertMethodCallContainsPattern('echo', 'Cannot use Docker credentials outside of jenkins infra environments'))
```

**Review by lemeurherve — APPROVED** [review]

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 9]
```diff
@@ -27,4 +41,38 @@ class InfraConfig implements Serializable {
...
+  def getDockerPushOrgAndCredentialsId() {
+    switch(jenkinsHostname){
+      case "ci.jenkins.io": 
+        return [error: false, organisation: "jenkins4eval", credentialId: "jenkins4eval-dockerhub-push"]
+        break;
+      case "trusted.ci.jenkins.io": 
+        return [error: false, organisation: "jenkins", credentialId: "jenkins-dockerhub-push"]
+        break;
+      case "infra.ci.jenkins.io": 
+        return [error: false, organisation: "jenkinsciinfra", credentialId: "jenkinsciinfra-dockerhub-push"]
```
Comment (dduportal) [inline src/io/jenkins/infra/InfraConfig.groovy:72]:
```suggestion
        return [error: false, organisation: "jenkinsinfraadmin", credentialId: "jenkinsciinfra-dockerhub-push"]
```

As we discovered together, the account `jenkinsinfraadmin` is the owner of the organization `jenkinsciinfra` so we have to keep it.

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 10]
```diff
@@ -1,22 +1,36 @@
...
   }
 
-  Boolean isRunningOnJenkinsInfra() {
-    return jenkinsURL == 'https://ci.jenkins.io/' || isTrusted() || isInfra()
+  Boolean isJenkinsURLcontains(String search) {
+    return jenkinsURL.startsWith('https://' + search + '.jenkins.io/') || jenkinsURL.startsWith('https://' + search + '.jenkins.io:')
+  }
+
+  Boolean isCI() {
+    return isJenkinsURLcontains('ci')
```
Comment (dduportal) [inline src/io/jenkins/infra/InfraConfig.groovy:21]:
```suggestion
    return this.jenkinsHostname == 'ci.jenkins.io'
```

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 11]
```diff
@@ -1,22 +1,36 @@
...
+    if (this.jenkinsURL != '') {
+      this.jenkinsHostname = new URL(this.jenkinsURL).getHost()
+    }
   }
 
-  Boolean isRunningOnJenkinsInfra() {
-    return jenkinsURL == 'https://ci.jenkins.io/' || isTrusted() || isInfra()
+  Boolean isJenkinsURLcontains(String search) {
+    return jenkinsURL.startsWith('https://' + search + '.jenkins.io/') || jenkinsURL.startsWith('https://' + search + '.jenkins.io:')
+  }
```
Comment (dduportal) [inline src/io/jenkins/infra/InfraConfig.groovy:18]:
```suggestion
```

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 12]
```diff
@@ -1,22 +1,36 @@
...
+
+  Boolean isCI() {
+    return isJenkinsURLcontains('ci')
   }
 
   Boolean isTrusted() {
-    return jenkinsURL.startsWith('https://trusted.ci.jenkins.io')
+    return isJenkinsURLcontains('trusted.ci')
```
Comment (dduportal) [inline src/io/jenkins/infra/InfraConfig.groovy:25]:
```suggestion
    return this.jenkinsHostname == 'trusted.ci.jenkins.io'
```

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 13]
```diff
@@ -1,22 +1,36 @@
...
   }
 
   Boolean isInfra() {
-    return jenkinsURL.startsWith('https://infra.ci.jenkins.io')
+    return isJenkinsURLcontains('infra.ci')
```
Comment (dduportal) [inline src/io/jenkins/infra/InfraConfig.groovy:29]:
```suggestion
    return this.jenkinsHostname == 'infra.ci.jenkins.io'
```

**Review by dduportal — CHANGES_REQUESTED** [review]
Following the last change that we did in mob-prgramming on this PR (adding a new object attribute `jenkinsHostname`, please check the 4 suggestions that simplifies the code for `IsInfra()` etc.). These suggestions also removes the risk of not checking strictly.

**Comment by dduportal** [comment]
TODO:
- Help me to finish filing accounts in https://github.com/jenkins-infra/runbooks/pull/47 
- Finish the PR based on this

**Comment by basil** [comment]
My apologies in advance for what must be a particularly harsh merge conflict. Let me know if you want me to resolve it for you.

**Comment by dduportal** [comment]
> My apologies in advance for what must be a particularly harsh merge conflict. Let me know if you want me to resolve it for you.

Many thanks for the notifications and the work! No worries we'll handle this: that would be a great practical exercise with git conflicts.

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 14]
```diff
@@ -27,4 +41,41 @@ class InfraConfig implements Serializable {
...
+        return [error: false, organisation: "trustedcijenkinsio", credentialId: "trustedcijenkinsio-dockerhub-pull"]
+        break
+      case "infra.ci.jenkins.io":
+        return [error: false, organisation: "infracijenkinsio", credentialId: "infracijenkinsio-dockerhub-pull"]
+        break
+      case "release.ci.jenkins.io":
+        return [error: false, organisation: "releasecijenkinsio", credentialId: "releasecijenkinsio-dockerhub-pull"]
+        break
+      default:
+        return [error: true, msg: "Cannot use Docker credentials outside of jenkins infra environments"]
```
Comment (dduportal) [inline src/io/jenkins/infra/InfraConfig.groovy:61]: TODO: use simple quotes for strings as there are no need for interpolation

File: src/io/jenkins/infra/InfraConfig.groovy [hunk 15]
```diff
@@ -27,4 +41,41 @@ class InfraConfig implements Serializable {
...
+        return [error: false, organisation: "jenkins4eval", credentialId: "cijenkinsio-dockerhub-push"]
+        break
+      case "trusted.ci.jenkins.io":
+        return [error: false, organisation: "jenkins", credentialId: "jenkinsciinfra-dockerhub-push"]
+        break
+      case "infra.ci.jenkins.io":
+        return [error: false, organisation: "jenkinsciinfra", credentialId: "jenkinsinfraadmin-dockerhub-push"]
+        break
+      default:
+        return [error: true, msg: "Cannot use Docker credentials outside of jenkins infra environments"]
```
Comment (dduportal) [inline src/io/jenkins/infra/InfraConfig.groovy:78]: TODO: use simple quotes for strings as there are no need for interpolation

**Review by dduportal — APPROVED** [review]
